File: app/campo.rb
#!/usr/bin/env ruby

# Copyright 2012 Welliton de Souza
#
# Este programa é um software livre; você pode redistribui-lo e/ou 
# modifica-lo dentro dos termos da Licença Pública Geral GNU como 
# publicada pela Fundação do Software Livre (FSF); na versão 2 da 
# Licença, ou (na sua opnião) qualquer versão.
#
# Este programa é distribuido na esperança que possa ser  util, 
# mas SEM NENHUMA GARANTIA; sem uma garantia implicita de ADEQUAÇÂO a qualquer
# MERCADO ou APLICAÇÃO EM PARTICULAR. Veja a
# Licença Pública Geral GNU para maiores detalhes.
#
# Você deve ter recebido uma cópia da Licença Pública Geral GNU
# junto com este programa, se não, escreva para a Fundação do Software
# Livre(FSF) Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

class Campo
  attr_accessor :grade
  
  def initialize(grade)
    @grade = grade
  end
  
  def posicionar_navio(navio, coordenada)
    orientacao = coordenada.orientacao
    if orientacao.eql? :horizontal
      navio.tamanho.times do |i|
        @grade[coordenada.x][coordenada.y + i].parte_navio = navio.partes[i]
      end
    elsif orientacao.eql? :vertical
      navio.tamanho.times do |i|
        @grade[coordenada.x + i][coordenada.y].parte_navio = navio.partes[i]
      end
    else
      raise "Coordenada invalida."
    end
  end
  
  def imprimir_rival
    saida = "\n"
    header = "     "  
    count = 1
    @grade.each do |x|
      saida << "%2d " % count
      header << "#{Coordenada.num2char(count-1)}   "
      count += 1
      x.each do |y|
        c = nil
        if y.atacado?
          if y.parte_navio.nil?
            c = "\e[34mO\e[0m"
          else
            c = "\e[31mX\e[0m"
          end
        else
          c = ' '
        end
        saida << "| #{c} "
      end
      saida << "|\n"
    end
    header << saida
  end
  
  def imprimir
    saida = "\n"
    header = "     "
    count = 1
    @grade.each do |x|
      saida << "%2d " % count
      header << "#{Coordenada.num2char(count-1)}   "
      count += 1
      x.each do |y|
        c = nil
        if y.parte_navio.nil?
          if y.atacado?
            c = " \e[34mO\e[0m "
          else
            c = '   '
          end
        else
          if y.atacado?
            c = "\e[31m[X]\e[0m"
          else
            c = '[ ]'
          end
        end
        saida << "|#{c}"
      end
      saida << "|\n"
    end
    header << saida
  end
  
  def atacar!(coordenada)
    if coordenada.is_a? Coordenada
      c = coordenada
    else
      c = Coordenada.parse coordenada
    end
    quadrante = @grade[c.x][c.y]
    quadrante.atacar!
    mensagem = nil
    if quadrante.parte_navio.nil?
      mensagem = "Nenhum navio foi atingido durante o ataque!\n"
    else
      mensagem = "Um navio inimigo foi atingido durante o ataque!\n"
      navio = quadrante.parte_navio.navio
      if navio.destruido?
        mensagem << "O navio #{navio.nome} foi destruido!\n"
      end
    end
    mensagem
  end
end

File: java_old/BatalhaNaval/src/batalhanaval/Campo.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package batalhanaval;

//import batalhanaval.Quadrante;
import java.util.ArrayList;

/**
 *
 * @author Larissa
 */
public class Campo {

    private Quadrante grade[][];
    private ArrayList<Navio> navios;

    public Campo(Quadrante grade[][]) {
        this.grade = grade;
        this.navios = new ArrayList<Navio>();
    }
    //Padrão de projeto expert information 

    public void posicionarNavio(Navio navio, Coordenada coordenada) {

        int x = coordenada.getX();
        int y = coordenada.getY();
        ArrayList<ParteNavio> partes = navio.getPartes();
        if (coordenada.getOrientacaoEnum().equals(OrientacaoEnum.HORIZONTAL)) {
            for (int i = 0; i < navio.getTamanho(); i++) {
                grade[x][y + i].setOcupado(partes.get(i));
            }
        } else if (coordenada.getOrientacaoEnum().equals(OrientacaoEnum.VERTICAL)) {
            for (int i = 0; i < navio.getTamanho(); i++) {
                grade[x + i][y].setOcupado(partes.get(i));
            }
        }
        getNavios().add(navio);
    }

    public String atacar(int x, int y) {
        String mensagem = "";
        Quadrante quadrante = grade[x][y];
        quadrante.setAtacado(true);
        ParteNavio parteNavio = quadrante.getOcupado();
        if (parteNavio != null) {
            mensagem = "Um Navio inimigo foi atingido durante o ataque";
            Navio navio = parteNavio.getPertence();
            if (navio.getDestruido()) {
                mensagem = "O navio " + navio.getNome() + " foi destruido";
            }
        } else {
            mensagem = "Nenhum navio foi atingido durante o ataque";
        }
        return mensagem;
    }

    public ArrayList<Navio> getNavios() {
        return navios;
    }

    public void setNavios(ArrayList<Navio> navios) {
        this.navios = navios;
    }

    
    public String imprimirCampo() {
        String saida = "\n";
        String header = "     ";
        int count = 1;
        for (int i = 0; i < grade.length; ++i) {
            saida += String.format("%2d ", count);
            header += String.format("%s   ", Coordenada.num2char(count - 1));
            count += 1;
            for (int j = 0; j < grade.length; ++j) {
                String c;
                Quadrante q = grade[i][j];
                if (q.getOcupado() == null) {
                    if (q.isAtacado()) {
                        c = " O ";
                    } else {
                        c = "   ";
                    }
                } else {
                    if (q.isAtacado()) {
                        c = "[X]";
                    } else {
                        c = "[ ]";
                    }
                }
                saida += String.format("|%s", c);
            }
            saida += "|\n";
        }
        return header + saida;
    }

    public String imprimirRival() {
        String saida = "\n";
        String header = "     ";
        int count = 1;
        for (int i = 0; i < grade.length; ++i) {
            saida += String.format("%2d ", count);
            header += String.format("%s   ", Coordenada.num2char(count - 1));
            count += 1;
            for (int j = 0; j < grade.length; ++j) {
                String c;
                Quadrante q = grade[i][j];
                if (q.isAtacado()) {
                    if (q.getOcupado() == null) {
                        c = "O";
                    } else {
                        c = "X";
                    }
                } else {
                    c = " ";
                }
                saida += String.format("| %s ", c);
            }
            saida += "|\n";
        }
        return header + saida;
    }
    
    public Quadrante[][] getGrade() {
        return grade;
    }
}
